src/debugpy/server/tracing/__init__.py

```diff
@@ -0,0 +1,533 @@
+        invoke directly; use StackFrame.from_frame_object() instead.
+        """
+        StackFrame._last_id += 1
+        self.id = StackFrame._last_id
+        self.thread = thread
+        self.frame_object = frame_object
+        self._path = None
+        self._scopes = None
+        self._all[self.id] = self
+
+    def __getstate__(self):
+        return {
+            "id": self.id,
+            "name": self.frame_object.f_code.co_name,
+            "source": {
+                # TODO: use "sourceReference" when path isn't available (e.g. decompiled code)
+                "path": str(self.path()),
+            },
+            "line": self.frame_object.f_lineno,
+            "column": 1,  # TODO
+            # TODO: "endLine", "endColumn", "moduleId", "instructionPointerReference"
+        }
+
+    @property
+    def line(self) -> int:
+        return self.frame_object.f_lineno
+
+    def path(self) -> Path:
+        if self._path is None:
+            path = Path(self.frame_object.f_code.co_filename)
+            try:
+                path = path.resolve()
+            except (OSError, RuntimeError):
+                pass
+            # No need to sync this since all instances are equivalent.
+            self._path = path
+        return self._path
+
+    def is_internal(self) -> bool:
+        # TODO: filter internal frames properly
+        parts = self.path().parts
+        internals = ["debugpy", "threading"]
+        return any(part.startswith(s) for s in internals for part in parts)
+
+    @classmethod
+    def from_frame_object(
+        self, thread: Thread, frame_object: FrameType
+    ) -> "StackFrame":
+        for frame in self._all.values():
+            if frame.thread is thread and frame.frame_object is frame_object:
+                return frame
+        return StackFrame(thread, frame_object)
+
+    @classmethod
+    def get(self, id: int) -> "StackFrame":
+        return self._all.get(id, None)
+
+    def scopes(self) -> List[Scope]:
+        if self._scopes is None:
+            self._scopes = [
+                Scope(self.frame_object, "local"),
+                Scope(self.frame_object, "global"),
+            ]
+        return self._scopes
+
+    @classmethod
+    def invalidate(self, thread_id: int):
+        frames = [frame for frame in self._all.values() if frame.thread.id == thread_id]
+        VariableContainer.invalidate(*frames)
+
+
+@dataclass
+class Step:
+    step: Literal["in", "out", "over"]
+    origin: FrameType = None
+    origin_line: int = None
+
+
+class Condition:
+    """
+    Expression that must be true for the breakpoint to be triggered.
+    """
+
+    expression: str
+    """Python expression that must evaluate to True for the breakpoint to be triggered."""
+
+    _code: CodeType
+
+    def __init__(self, breakpoint: "Breakpoint", expression: str):
+        self.expression = expression
+        self._code = compile(
```

Comment:
Should we have a syntax error catch here and re-raise with a message that the syntax error was in the condition? Otherwise, it might look like a generic syntax error.

Reply:
You're right, it should catch. But it can't re-raise because that would crash the debuggee. Intuitively it feels like this should be an error response to the "breakpoints" request so that the user immediately gets an error, and this is what I implemented, but VSCode doesn't seem to be propagating that failure to the user in practice. I'll investigate further, but let's track it as a separate issue.